### PR TITLE
Fix recompiles of stale dependencies and cyclic graphs

### PR DESCRIPTION
**Pull request by alexcrichton**
Set out to solve #1236, but ended up solving #834 as well!

### CODE REVIEW

File: src/cargo/core/resolver/mod.rs [hunk 1]
```diff
@@ -131,28 +131,45 @@ struct Context {
...
     }
 }
 
 fn activate(mut cx: Box<Context>,
             registry: &mut Registry,
-            parent: &Summary,
+            parent: &Rc<Summary>,
             method: Method)
             -> CargoResult<CargoResult<Box<Context>>> {
+    // Dependency graphs are required to be a DAG. Non-transitive
```
Comment (huonw) [inline src/cargo/core/resolver/mod.rs]: This comment makes less sense after the move.

Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: Good point, updated.

**Pull request merged**